### PR TITLE
added mixed indexing and bounds checking

### LINKED ISSUE #75: Support mixed indexing (lists and slices)
We should be able to use mixed list and slice indexing, for example,

```python
barray = bolt.array(arange(24).reshape(2,3,4), context=sc)
barray[:,[0,2],:]
```
which currently throws a `NotImplementedError`

for reference, it should return exactly the same thing as numpy, e.g.

```python
a = barray[:,[0,2],:].toarray()
b = barray.toarray()[:,[0,2],:]
assert allclose(a, b)
```

thanks to @boazmohar for flagging, cc @jwittenbach

### PR DESCRIPTION
**Pull request by jwittenbach**
Fixes #75 and #76

Mixed indexing is implemented, but only for the restricted case where only one index uses advanced indexing, while all of the others use basic indexing.

Also adds bounds-checking on indexing so that errors are caught during lazy evaluation. This is implemented such that indexing will either behave identically to NumPy or raise an error if we do not support the particular indexing pattern. Right now there are two such main patterns:
1. The above-mentioned mixed indexing with more than one advanced index
2. Any slicing operation that, in NumPy, would result in an empty dimension (size = 0), as the `BoltArraySpark` is not particularly well-equipped to handle this case at the moment.

### CODE REVIEW

File: bolt/spark/array.py [hunk 1]
```diff
@@ -500,6 +500,41 @@ def key_func(key):
...
+        index = list(index)
+        loc = where([isinstance(i, (tuple, list, ndarray)) for i in index])[0][0]
+        idx = list(index[loc])
+
+        if isinstance(idx[0], (tuple, list, ndarray)):
+            raise ValueError("When mixing basic and advanced indexing, advanced index must be one-dimensional")
+
+        # handle the single advanced index
+        if loc < self.split:    # advanced index is on a key -- filter and update key
+            def newkey(key):
```
Comment (freeman-lab) [inline bolt/spark/array.py]: i prefer to not use same line comments, and rather reword so the comment goes above the `if...else...` block

File: test/spark/test_spark_getting.py [hunk 2]
```diff
@@ -79,6 +80,51 @@ def test_getitem_list_array(sc):
...
+    i = [[0, 1], [1, 0]]
+    with pytest.raises(ValueError):
+        b[i, :, :, :]
+
+def test_bounds(sc):
+
+    x = arange(5)
+    b = array(x, sc)
+
+    # out of bounds
```
Comment (freeman-lab) [inline test/spark/test_spark_getting.py]: Very nice, now that I'm seeing this, would love to hear someone explain why `numpy` doesn't do this itself... there must be a reason, but this feels like a cleaner API.

**Comment by freeman-lab** [comment]
Really nice job on this @jwittenbach ! No major comments, just one style nit.

File: test/spark/test_spark_getting.py (hunk 2 above)
Comment (jwittenbach) [inline test/spark/test_spark_getting.py]: Completely agree -- it would have helped me solve any number of off-by-one errors way more quickly. It does allow for some hacks, but a clean API should need hacks on strange edge-cases!

**Comment by jwittenbach** [comment]
Reformatted the comments and all tests still pass (bad news if they didn't :wink:). Merging!
